groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -0,0 +1,1653 @@
+    // trivially destructible types, and it provides a destructor that ensures
+    // the value_type destructor is called if 'd_buffer' holds an object.  The
+    // specialisation for 'is_trivially_destructible' types does not have a
+    // user provided destructor and 'is_trivially_destructible' itself.
+
+  public:
+    // CREATORS
+    ~Optional_Data();
+        // Destroy the managed 'value_type' object, if it exists.
+
+};
+
+                        // ===================
+                        // class Optional_Data
+                        // ===================
+
+template <class TYPE>
+struct Optional_Data<TYPE, true> : public Optional_DataImp<TYPE> {
+    // Component-level class; DO NOT USE
+
+    // Optional_Data is a type used to manage 'value_type' object in
+    // 'optional'.  It allows 'optional<TYPE>' to be trivially destructible if
+    // 'TYPE' is trivially destructible.  This class also provides an
+    // abstraction for const value type.  An 'optional' may contain a 'const'
+    // type object.  An assignment to such an 'optional' should not succeed.
+    // However, unless the 'optional' itself is 'const', it should be possible
+    // to change the value of the 'optional' using 'emplace'.  In order to
+    // allow for that, this class manages a non-const object of 'value_type',
+    // but all the observers return a 'const' adjusted reference to the managed
+    // object.  This functionality is common for all value types and is
+    // implemented in the Optional_DataImp base class.  The derived class,
+    // 'Optional_Data', is specialised on 'value_type'
+    // 'is_trivially_destructible' trait.  The main template is for non
+    // trivially destructible types, and it provides a destructor that ensures
+    // the value_type destructor is called if 'd_buffer' holds an object.  This
+    // specialisation is for 'is_trivially_destructible' types, and does not
+    // have a user provided destructor, which makes it
+    // 'is_trivially_destructible' itself.
+
+  public:
+
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+    // workaround for C++03 bsl::is_trivially_copyable trait.  Note that,
+    // whether 'Optional_Data<TYPE>' satisfies 'bsl::is_trivally_copyable'
+    // doesn't affect 'Optional<TYPE>' 'bsl::is_trivally_copyable' trait.  We
+    // only add this nested trait for the tests to be able to check the C++03
+    // implementation of 'Optional_Data'.  For correct C++03 functionality,
+    // 'bsl::optional' has to add a nested trait as well.
+    BSLMF_NESTED_TRAIT_DECLARATION_IF(Optional_Data,
+                           bsl::is_trivially_copyable,
+                           bsl::is_trivially_copyable<TYPE>::value);
+#endif //BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+
+};
+
+
+template <class TYPE, bool IS_TRIVIALLY_DESTRUCTIBLE>
+Optional_Data<TYPE, IS_TRIVIALLY_DESTRUCTIBLE>::~Optional_Data()
+{
+    this->reset();
+}
```

Comment:
This also goes under `INLINE DEFINITIONS`  banner, under its own class banner.

Reply:
done